pkg/controller/constrainttemplate/constrainttemplate_controller_test.go

```diff
@@ -672,48 +756,88 @@ func makeCRD(gvk schema.GroupVersionKind, plural string) *apiextensionsv1.Custom
 }
 
 // applyCRD applies a CRD and waits for it to register successfully.
-func applyCRD(ctx context.Context, g *gomega.GomegaWithT, client client.Client, gvk schema.GroupVersionKind, crd client.Object) error {
+func applyCRD(ctx context.Context, client client.Client, gvk schema.GroupVersionKind, crd client.Object) error {
 	err := client.Create(ctx, crd)
 	if err != nil {
 		return err
 	}
 
 	u := &unstructured.UnstructuredList{}
 	u.SetGroupVersionKind(gvk)
-	g.Eventually(func() error {
+	return retry.OnError(constantRetry, func(err error) bool {
+		return true
+	}, func() error {
 		if ctx.Err() != nil {
 			return ctx.Err()
 		}
 		return client.List(ctx, u)
-	}, 5*time.Second, 100*time.Millisecond).ShouldNot(gomega.HaveOccurred())
-	return nil
+	})
 }
 
-func deleteObjectAndConfirm(ctx context.Context, c client.Client, obj client.Object, timeout time.Duration) error {
-	key := client.ObjectKeyFromObject(obj)
+func deleteObjectAndConfirm(ctx context.Context, t *testing.T, c client.Client, obj client.Object) func() {
+	t.Helper()
 
-	err := c.Delete(ctx, obj)
-	if apierrors.IsNotFound(err) {
-		return nil
-	} else if err != nil {
-		return err
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	namespace := obj.GetNamespace()
+	name := obj.GetName()
+	if gvk.Empty() {
+		t.Fatalf("gvk for %v/%v %T is empty",
+			namespace, name, obj)
 	}
 
-	err = wait.Poll(10*time.Millisecond, timeout, func() (bool, error) {
-		// Get the object name
-		err2 := c.Get(ctx, key, obj)
-		if err2 != nil {
+	return func() {
+		t.Helper()
+
+		toDelete := makeUnstructured(gvk, namespace, name)
+		err := c.Delete(ctx, toDelete)
+		if apierrors.IsNotFound(err) {
+			return
+		} else if err != nil {
+			t.Fatal(err)
+		}
+
+		err = retry.OnError(constantRetry, func(err error) bool {
+			return true
+		}, func() error {
+			// Get the object name
+			toGet := makeUnstructured(gvk, namespace, name)
+			key := client.ObjectKey{Namespace: namespace, Name: name}
+			err2 := c.Get(ctx, key, toGet)
 			if apierrors.IsGone(err2) || apierrors.IsNotFound(err2) {
-				return true, nil
+				return nil
 			}
-			return false, err2
+
+			s, _ := json.MarshalIndent(toGet, "", "  ")
+			return fmt.Errorf("found %v %v:\n%s", gvk, key, string(s))
+		})
+
+		if err != nil {
+			t.Fatal(err)
 		}
-		return false, err2
-	})
-	return err
+	}
 }
 
 // This interface is getting used by tests to check the private objects of objectTracker.
 type testExpectations interface {
 	IsExpecting(gvk schema.GroupVersionKind, nsName types.NamespacedName) bool
 }
+
+func createThenCleanup(ctx context.Context, t *testing.T, c client.Client, obj client.Object) {
+	t.Helper()
+	err := c.Create(ctx, obj.DeepCopyObject().(client.Object))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Cleanup(deleteObjectAndConfirm(ctx, t, c, obj))
```

Comment:
should obj be deepcopied here to avoid obj getting overwritten?

Reply:
It is unnecessary to do so. I've added explanatory comments.